lessons/specifics/guardian.md

```diff
@@ -0,0 +1,264 @@
+
+You do not need to track JWT via a database. You can simply rely on the issued and expiry timestamps for controlling access. Often you'll end up using a database to look up your user resource but the JWT itself does not require it.
+
+For example, if you were going to use JWT to authenticate communication on a UDP socket you likely wouldn't use a database. Encode all the information you need directly into the token when you issue it. Once you verify it (check that it's signed correctly) you're good to go.
+
+You _can_ however use a database to track JWT. If you do, you gain the ability to verify that the token is still valid - that is - it has not been revoked. Or you could use the records in the db to force a logout of all tokens for user 5. This is made simple in Guardian by using [GuardianDb](https://github.com/hassox/guardian_db). GuardianDb uses Guardians 'Hooks' to perform validation checks, save and delete from the db. We'll cover that later.
+
+### Can I use someone elses token?
+
+Often folks wonder if it's appropriate to use an OAuth token from Facebook or Google as their authentication token. It is not. The behaviours of those tokens, their validity is defined outside your application. There's a more security focused reason though. Anyone who obtains the token not only has access to your site, they also have access to Facebook/Google/Some Provider. 
+
+Always use your own tokens, other applications tokens are useful to identify someone and make requests against their systems.
+
+## Setup
+
+There are many options for setting up Guardian. We'll cover them at some point but lets start with a very simple setup.
+
+### Minimal Setup
+
+To get started there are a handlful of things that you'll need. 
+
+#### Configuration
+
+`mix.exs`
+
+```elixir
+def application do
+  [
+    mod: {MyApp, []},
+    applications: [:guardian, ...]
+  ]
+end
+
+def deps do
+  [
+    {guardian: "~> x.x"},
+    ...
+  ]
+end
+```
+
+`config/config.ex`
+
+```elixir
+config :guardian, Guardian,
+  issuer: "MyAppId",
+  secret_key: Mix.env, # in each environment config file you should overwrite this if it's external
+  serializer: MyApp.GuardianSerializer
+```
+
+This is the minimum set of information you need to provide Guardian with to operate. You shouldn't encode your secret key directly into your top level config. Instead each environment should have it's own key. I tend to use the Mix environment for secrets in dev and test. Staging and production however I use secrets generated with `mix phoenix.gen.secret`. 
+
+`lib/my_app/guardian_serializer.ex`
+
+```elixir
+defmodule MyApp.GuardianSerializer do
+  @behaviour Guardian.Serializer
+
+  alias MyApp.Repo
+  alias MyApp.User
+
+  def for_token(user = %User{}), do: { :ok, "User:#{user.id}" }
+  def for_token(_), do: { :error, "Unknown resource type" }
+
+  def from_token("User:" <> id), do: { :ok, Repo.get(User, id) }
+  def from_token(_), do: { :error, "Unknown resource type" }
+end
+```
```

Comment:
There's not actually a _need_ for a user. In this case I'm using a User model because that's the most common case. I also have apps where is just use a simple string and don't have any model associated with it. I'll have a think about some words to that effect.